### PR TITLE
Internal PR for testing #866

### PR DESCRIPTION
**Pull request by meiao**
This is an internal PR to run the tests for the external PR #866

### CODE REVIEW

File: instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java [hunk 1]
```diff
@@ -59,55 +33,62 @@ public static void recordNewRelicLogEvent(LogEvent event) {
             Throwable throwable = event.getThrown();
 
             if (shouldCreateLogEvent(message, throwable)) {
-                HashMap<String, Object> logEventMap = new HashMap<>(DEFAULT_NUM_OF_LOG_EVENT_ATTRIBUTES);
+                Map<LogAttributeKey, Object> logEventMap = new HashMap<>(DEFAULT_NUM_OF_LOG_EVENT_ATTRIBUTES);
                 if (message != null) {
                     String formattedMessage = message.getFormattedMessage();
                     if (formattedMessage != null && !formattedMessage.isEmpty()) {
-                        logEventMap.put(MESSAGE, formattedMessage);
+                        logEventMap.put(AppLoggingUtils.MESSAGE, formattedMessage);
```
Comment (jasonjkeller) [inline instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java:40]: Do you think it's worth adding a static import of `AppLoggingUtils` to avoid having to use the `AppLoggingUtils.` prefix all over?

File: instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java [hunk 2]
```diff
@@ -59,55 +33,62 @@ public static void recordNewRelicLogEvent(LogEvent event) {
...
                     }
                 }
-                logEventMap.put(TIMESTAMP, event.getTimeMillis());
+                logEventMap.put(AppLoggingUtils.TIMESTAMP, event.getTimeMillis());
+
+                if (AppLoggingUtils.isAppLoggingContextDataEnabled() && event.getContextData() != null) {
+                    event.getContextData().forEach((key, value) -> {
```
Comment (jasonjkeller) [inline instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java:46]: Is this implementation of `forEach` just a simple for loop under the hood?

If not, isn't this much less performant that using a simple loop? For the extra allocation reasons that you pointed out previously? Also based on some experimentation that I've done I found that the execution time is much longer when using streaming operations on small data sets as compared to a simple loop. Unless you expect that there may be many thousands of context attributes I'm not sure of the benefit of streaming.

File: instrumentation/spring-aop-2/build.gradle [hunk 3]
```diff
@@ -4,9 +4,6 @@ dependencies {
     implementation("org.springframework:spring-aop:3.2.+")
     //     implementation("org.springframework:spring:2.0.3") // First version to have MethodMatchers.matches()
 
-    testImplementation("org.powermock:powermock-module-junit4-rule-agent:1.6.1")
-    testImplementation("org.powermock:powermock-module-junit4:1.6.1")
-    testImplementation("org.powermock:powermock-api-mockito:1.6.1")
```
Comment (jasonjkeller) [inline instrumentation/spring-aop-2/build.gradle:9]: How is this related to this PR?

File: newrelic-agent/src/main/java/com/newrelic/agent/attributes/AttributesNode.java [hunk 4]
```diff
@@ -79,7 +79,7 @@ protected Boolean applyRules(String key) {
     }
 
     private void logMatch(String key) {
-        if (Agent.LOG.isFinerEnabled()) {
+        if (Agent.LOG.isFinestEnabled()) {
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/java/com/newrelic/agent/attributes/AttributesNode.java:82]: doh!

File: newrelic-agent/src/main/resources/newrelic.yml [hunk 5]
```diff
@@ -105,6 +105,12 @@ common: &default_settings
       # Default is 10000. Setting to 0 will disable.
       #max_samples_stored: 10000
 
+      # Whether the log events should include MDC from loggers with support for that.
+      include_context_data:
+
+        # When true, application logs will contain context data.
+        enabled: false
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/resources/newrelic.yml:112]: I wonder if we should also put the `include`/`exclude` options in the default yaml but keep them commented out like we do elsewhere?

```
      # A comma separated list of attribute keys whose values should be sent to New Relic.
      #include:

      # A comma separated list of attribute keys whose values should not be sent to New Relic.
      #exclude:
```

**Review by jasonjkeller — COMMENTED** [review]
Overall looks good! Just a couple of questions and potential cleanup things.

I like the cleanup introduced with the `AppLoggingUtils`, that's a nice quality of life enhancement.

File: instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java (hunk 2 above)
Comment (meiao) [reply instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java:46]: This is a bit of a different case. `event.getContextData()` does not return a Map. It returns a ReadOnlyStringMap which extends Serializable and implements no interfaces.
This ReadOnlyStringMap provides a toMap method and the forEach that can be used here. There are many implementation of that interface, but the one I got in a unit test was SortedArrayStringMap. This impl, will instantiate a new HashMap when toMap is called. The forEach implementation is done using a for(int i...) loop.
So I'd have to do some profiling to see which one allocates less memory.

File: instrumentation/spring-aop-2/build.gradle (hunk 3 above)
Comment (meiao) [reply instrumentation/spring-aop-2/build.gradle:9]: I was looking for mocking static methods. Found these and noticed that this module had no tests.

**Review by jasonjkeller — APPROVED** [review]
LGTM!!!
